nva/nva.c

```diff
@@ -135,6 +147,84 @@ struct nva_card *nva_init_apu(struct pci_device *dev) {
 	return card;
 }
 
+static int check_modalias(const char *node_name) {
+	char path[PATH_MAX];
+	char buffer[255] = { 0 };
+	FILE *file;
+
+	/* Check if the GPU is indeed made by NVIDIA by checking the compatibility
+	 * field of the device tree, exposed to the userspace via modalias. As
+	 * we are lazy, let's not parse the modalias string but just check if
+	 * it mentions nvidia at all. If it does, return 0.
+	 */
+
+	snprintf(path, PATH_MAX, "%s/%s/modalias", PLATFORM_DEVICES_DIR, node_name);
+	if (!(file = fopen(path, "r"))) {
+		return 2;
+	}
+	fread(buffer, 1, sizeof(buffer), file);
+	return strstr(buffer, "nvidia") == NULL;
+}
+
+struct nva_card *nva_init_platform() {
+	uint32_t tmp, bar0 = 0xffffffff;
+	struct nva_card *card;
+	struct dirent *dent;
+	DIR *dir;
+	int fd;
+
+	/* Look for a gpu on the "platform" bus by looking at the platform bus,
+	 * looking for an entry such as '/sys/bus/platform/devices/57000000.gpu'.
+	 * We only expect up to one entry.
+	 */
+	if (!(dir = opendir(PLATFORM_DEVICES_DIR)))
+		return NULL;
+	while ((dent = readdir(dir))) {
+		/* FIXME: this may not work on downstream kernels */
+		const char *last_four = dent->d_name + strlen(dent->d_name) - 4;
+		if (strcmp(last_four, ".gpu") == 0) {
+			if (sscanf(dent->d_name, "%x.gpu", &tmp) == 1) {
+				/* Check if the modalias contains nvidia */
+				if (check_modalias(dent->d_name))
+					continue;
+
+				/* We have found the gpu, exit the loop */
+				bar0 = tmp;
+				break;
+			}
+		}
+	}
+
+	if (bar0 == 0xffffffff)
+		return NULL;
+
+	/* we have found a gpu */
+	card = calloc(sizeof *card, 1);
+	if (!card)
+                return 0;
+	card->type = NVA_DEVICE_GPU;
+	card->bus_type = NVA_BUS_PLATFORM;
+	card->bus.platform_address = bar0;
+
+	if((fd = open("/dev/mem", O_RDWR | O_SYNC)) == -1) {
+		fprintf (stderr, "WARN: Can't open /dev/mem. Are you root?\n");
+		free(card);
+		return 0;
+	}
+
+	card->bar0 = mmap(0, 0x1000000, PROT_READ | PROT_WRITE, MAP_SHARED, fd, bar0);
```

Comment:
I do not think there is anything like this when using the platform bus. For your information, it is a non-discoverable bus (which made Linus bitch pretty hard on it, reference included at the end of this message). The only address that we have is what the device tree tells us, which is 0x5700000 or something like that. I called it bar0 because it is containing what we would find on pcie devices but I do not think there are VRAM appertures and bar3 (what does it contain again?) in the tegra world.

Linus:
Ok. I still really despise the absolute incredible sh*t that is non-discoverable buses, and I hope that ARM SoC hardware designers all die in some incredibly painful accident.  DT only does so much. So if you see any, send them my love, and possibly puncture the brake-lines on their car and put a little surprise in their coffee, ok?